lib/str.d.ts

```diff
@@ -0,0 +1,643 @@
+    toBool(value: unknown): boolean;
+    /**
+     * Checks if a string could be the masked version of another one.
+     *
+     * @param {String} first string to compare
+     * @param {String} second string to compare
+     * @param {String} [mask] defaults to X
+     * @return {Boolean} true when first could be the masked version of second
+     */
+    maskedEquals(first: string, second: string, mask: string): boolean;
+    /**
+     * Bold any word matching the regexp in the text.
+     * @param {string} text, htmlEncoded
+     * @param {RegExp} regexp
+     * @return {string}
+     */
+    boldify(text: string, regexp: RegExp): string;
+    /**
+     * Check for whether a phone number is valid.
+     * @param {String} phone
+     *
+     * @return {bool}
+     */
+    isValidPhone(phone: string): boolean;
+    /**
+     * We validate mentions by checking if it's first character is an allowed character.
+     *
+     * @param {String} mention
+     * @returns {bool}
+     */
+    isValidMention(mention: string): boolean;
+    /**
+     * Returns text without our SMS domain
+     *
+     * @param {String} text
+     * @return {String}
+     */
+    removeSMSDomain(text: string): string;
+    /**
+     * Returns true if the text is a valid phone number with our SMS domain removed
+     *
+     * @param {String} text
+     * @return {bool}
+     */
+    isSMSLogin(text: string): boolean;
+    /**
+     * This method will return all matches of a single regex like preg_match_all() in PHP. This is not a common part of
+     * JS yet, so this is a good way of doing it according to
+     * https://github.com/airbnb/javascript/issues/1439#issuecomment-306297399 and doesn't get us in trouble with
+     * linting rules.
+     *
+     * @param {String} str
+     * @param {RegExp} regex
+     *
+     * @returns {Array}
+     */
+    matchAll(str: string, regex: RegExp): unknown[];
+    /**
+     * A simple GUID generator taken from https://stackoverflow.com/a/32760401/9114791
+     *
+     * @param {String} [prefix] an optional prefix to put in front of the guid
+     * @returns {String}
+     */
+    guid(prefix?: string): string;
+    /**
+     * Takes in a URL and returns it with a leading '/'
+     *
+     * @param {mixed} url The URL to be formatted
+     * @returns {String} The formatted URL
+     */
+    normalizeUrl(url: string): string;
+    /**
+     *  Formats a URL by converting the domain name to lowercase and adding the missing 'https://' protocol.
+     *
+     * @param {url} url The URL to be formatted
+     * @returns {String} The formatted URL
+     */
+    sanitizeURL(url: string): string;
+    /**
+     * Checks if parameter is a string or function
+     * if it is a function then we will call it with
+     * any additional arguments.
+     *
+     * @param {String|Function} parameter
+     * @returns {String}
+     */
+    result<R, A>(
+        parameter: string | ((...args: A[]) => R),
+        ...args: A[]
+    ): string | R;
```

Comment:
`...args: A[]` means that all arguments have to be the same type of `A`, which is not the same for the `result` method.

I suggest the following code. This uses function overload and `extends` keyword to provide more information as well as specifying that arguments can be any number of arbitrary types.

```suggestion
    result: {
    (parameter: string): string;
    <R, A extends unknown[]>(parameter: (...args: A) => R, ...args: A): R;
  }
```

Reply:
Nice one! Adjusted 😄